<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Descripción-del-proyecto" data-toc-modified-id="Descripción-del-proyecto-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Descripción del proyecto</a></span><ul class="toc-item"><li><span><a href="#Descrpcion-de-los-datos" data-toc-modified-id="Descrpcion-de-los-datos-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Descrpcion de los datos</a></span></li><li><span><a href="#Diagrama-de-datos" data-toc-modified-id="Diagrama-de-datos-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Diagrama de datos</a></span></li></ul></li><li><span><a href="#Importación-de-librerías-y-conexión-a-la-base-de-datos" data-toc-modified-id="Importación-de-librerías-y-conexión-a-la-base-de-datos-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Importación de librerías y conexión a la base de datos</a></span></li><li><span><a href="#Estudio-de-tablas." data-toc-modified-id="Estudio-de-tablas.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Estudio de tablas.</a></span><ul class="toc-item"><li><span><a href="#Tabla-books" data-toc-modified-id="Tabla-books-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Tabla books</a></span></li><li><span><a href="#Tabla-authors" data-toc-modified-id="Tabla-authors-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Tabla authors</a></span></li><li><span><a href="#Tabla-publishers" data-toc-modified-id="Tabla-publishers-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Tabla publishers</a></span></li><li><span><a href="#Tabla-ratings" data-toc-modified-id="Tabla-ratings-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Tabla ratings</a></span></li><li><span><a href="#Tabla-review" data-toc-modified-id="Tabla-review-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Tabla review</a></span></li></ul></li><li><span><a href="#Ejercicios" data-toc-modified-id="Ejercicios-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ejercicios</a></span><ul class="toc-item"><li><span><a href="#Busco-número-de-libros-publicados-después-del-1-de-enero-de-2000." data-toc-modified-id="Busco-número-de-libros-publicados-después-del-1-de-enero-de-2000.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Busco número de libros publicados después del 1 de enero de 2000.</a></span></li><li><span><a href="#Busco-el-número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro." data-toc-modified-id="Busco-el-número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro.-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Busco el número de reseñas de usuarios y la calificación promedio para cada libro.</a></span></li><li><span><a href="#Busco-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas." data-toc-modified-id="Busco-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas.-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Busco la editorial que ha publicado el mayor número de libros con más de 50 páginas.</a></span></li><li><span><a href="#Identifico-al-autor-que-tiene-la-más-alta-calificación-promedio-:-miro-solo-los-libros-con-al-menos-50-calificaciones." data-toc-modified-id="Identifico-al-autor-que-tiene-la-más-alta-calificación-promedio-:-miro-solo-los-libros-con-al-menos-50-calificaciones.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Identifico al autor que tiene la más alta calificación promedio : miro solo los libros con al menos 50 calificaciones.</a></span></li><li><span><a href="#Encuentra-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros." data-toc-modified-id="Encuentra-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.</a></span></li></ul></li><li><span><a href="#Conclusiones" data-toc-modified-id="Conclusiones-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusiones</a></span></li></ul></div>

# Proyecto  final SQL

## Descripción del proyecto

El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina y hábitos diarios de las personas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera yendo a cafés y centros comerciales sino que comenzaron a desarrollar nuevas actividades en sus casas, entre ellas la lectura. Eso atrajo la atención de las startups que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

En este proyecto trabajaré con una base de datos de uno de los servicios que compiten en el mercado analizando su actividad. Este dataset, contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros.

### Descrpcion de los datos

Tabla **books**

Contiene datos sobre libros:

- `book_id`: identificación del libro
- `author_id`: identificación del autor o autora
- `title`: título
- `num_pages`: número de páginas
- `publication_date`: fecha de la publicación
- `publisher_id`: identificación de la editorial

Tabla **authors**

Contiene datos sobre autores:

- `author_id`: identificación del autor o autora
- `author`: el autor o la autora

Tabla **publishers**

Contiene datos sobre editoriales:

- `publisher_id`: identificación de la editorial
- `publisher`: la editorial

Tabla **ratings**

Contiene datos sobre las calificaciones de usuarios:

- `rating_id`: identificación de la calificación
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `rating`: calificación

Tabla **reviews**

Contiene datos sobre las reseñas de los y las clientes:

- `review_id`: identificación de la reseña
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `text`: el texto de la reseña

### Diagrama de datos


![der](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F5a5cff26-7d06-45cb-8333-37837bbd1a81%2FUntitled.png?id=e1f8ab17-67c0-49e7-9a68-bc20d5ce977c&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2)


A continuación comenzaré con la importacón de las librerías y conexión con la base de datos.
    
    
## Importación de librerías y conexión a la base de datos 

In [ ]:
# importar librerías

import pandas as pd
from sqlalchemy import create_engine

In [ ]:
# Conexión a la base de datos

db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Estudio de tablas.

A continuación procederé a estudiar las tablas de la base de datos.

### Tabla books

In [ ]:
# Estudio tabla books

query= ''' select *
           from books
           limit 5'''

books= pd.io.sql.read_sql(query, con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [ ]:
# Analizo tabla books

query= ''' select count(book_id) as number_of_books,
                  count(distinct author_id) as authors,
                  count(distinct title) as titles,
                  min(publication_date) as first_publication_date,
                  max(publication_date) as last_publication_date,
                  count(distinct publisher_id) as publishers
           from books
           '''

b1= pd.io.sql.read_sql(query, con = engine)
b1

number_of_books  authors  titles first_publication_date  \
0             1000      636     999             1952-12-01   

  last_publication_date  publishers  
0            2020-03-31         340

In [ ]:
# Busco título duplicado 

query= ''' select title,
                  count(title) as book_amount
           from books
           group by title
           having count(title) > 1
           '''

b2= pd.io.sql.read_sql(query, con = engine)
b2

title  book_amount
0  Memoirs of a Geisha            2

In [ ]:
# Busco más información del libro Memoirs of a Geisha

query= ''' select book_id,
                  title,
                  author_id,
                  publisher_id,
                  publication_date
           from books
           where title = 'Memoirs of a Geisha'
           group by book_id,
                    title,
                    author_id,
                    publisher_id,
                    publication_date
          
           '''

b3= pd.io.sql.read_sql(query, con = engine)
b3

book_id                title  author_id  publisher_id publication_date
0      426  Memoirs of a Geisha         39           241       2005-11-15
1      427  Memoirs of a Geisha         39           311       2005-11-22

De la tabla **books** podemos conocer que el servicio de lectura cuenta con una variedad de 999 títulos diferentes pertenecientes a 636 autores.

Las fechas de publicación de los ejemplares van desde 01-12-1952 hasta el 31-01-2020 pertenecientes a 340 editoriales distintas.

Podemos observar que el número de libros que aparece es de 1000 porque tenemos un libro (Memoirs of a Geisha) que fue publicado por dos editoriales diferentes en distinta fecha.


### Tabla authors

In [ ]:
# Estudio tabla authors

query= ''' select *
           from authors
           limit 5'''


authors = pd.io.sql.read_sql(query, con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
# Busco la cantidad de autores

query= ''' select count(author_id) as number_of_authors,
                  count(distinct authors) as authors
           from authors
           '''


a1 = pd.io.sql.read_sql(query, con = engine)
a1

number_of_authors  authors
0                636      636

La tabla **authors** muestra los nombres de los autores (636) de los libros.

### Tabla publishers

In [ ]:
# Estudio tabla publishers

query= ''' select *
           from publishers
           limit 5'''


publishers = pd.io.sql.read_sql(query, con = engine)
publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
# Busco cantidad de editoriales

query= ''' select count(publisher_id) as number_of_publishers
           from publishers
           '''


p1 = pd.io.sql.read_sql(query, con = engine)
p1

number_of_publishers
0                   340

La tabla **publisher** muestra los nombres de las 340 editoriales que publicaron los libros mencionados en la tabla **books**.

### Tabla ratings

In [ ]:
# Estudio tabla ratings

query= ''' select *
           from ratings
           limit 5'''


ratings = pd.io.sql.read_sql(query, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
# Analizo la tabla ratings

query= ''' select count(rating_id) as number_of_ratings,
                  count(distinct book_id) as rated_books,
                  count(distinct username) as number_of_users,
                  max(rating) as max_rating_number,
                  min(rating) as min_rating_number
           from ratings
           '''

r1 = pd.io.sql.read_sql(query, con = engine)
r1

number_of_ratings  rated_books  number_of_users  max_rating_number  \
0               6456         1000              160                  5   

   min_rating_number  
0                  1

La tabla ratings nos muestra que se hicieron 6456 calificaiones. Las mismas fueron realizadas por 160 personas distintas. Todos los libros presentes en la plataforma del servicio fueron evaluados. Estas calificaciones van desde 1 a 5. 

### Tabla review

In [ ]:
# Estudio tabla reviews

query= ''' select *
           from reviews
           limit 5'''


reviews = pd.io.sql.read_sql(query, con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [ ]:
# Analizo tabla reviews

query= ''' select count(review_id) as number_of_reviews,
                  count(distinct book_id) as books_reviwed,
                  count(distinct username) as reviewers
           from reviews
           '''


r1 = pd.io.sql.read_sql(query, con = engine)
r1

number_of_reviews  books_reviwed  reviewers
0               2793            994        160

La tabla reviews nos muestra que se realizaron 2793 reseñas sobre 994 libros. Esas reseñas fueron realizadas por 160 usuarios. De los 1000 libros presentes en el servicio, solo 6 no recibieron una reseña. 

## Ejercicios

### Busco número de libros publicados después del 1 de enero de 2000.

In [ ]:
# Encuentro el número de libros publicados después del 1 de enero de 2000.

query= '''select count(book_id) as cantidad_de_libros
          from books
          where publication_date > '2000-01-1'
'''


ej_1 = pd.io.sql.read_sql(query, con = engine)
ej_1

cantidad_de_libros
0                 819

Después del 1 de enero de 2000 se publicaron 819 libros en total.

### Busco el número de reseñas de usuarios y la calificación promedio para cada libro.

In [ ]:
# Encuentro el número de reseñas de usuarios y la calificación promedio para cada libro.

query= ''' select books.book_id,
                  authors.author,
                  books.title,
                  count(distinct reviews.review_id) as review,
                  round(avg(ratings.rating),2) as avg_ratings 
           from books
                left join authors on books.author_id = authors.author_id
                left join reviews on reviews.book_id = books.book_id
                left join ratings on ratings.book_id = books.book_id
          group by books.book_id, authors.author_id
          order by avg_ratings desc
'''


ej_2 = pd.io.sql.read_sql(query, con = engine)
ej_2.head()

book_id                                             author  \
0       86                                    Mercedes Lackey   
1      901  Robert Kirkman/Tony Moore/Charlie Adlard/Cliff...   
2      390                                   William Faulkner   
3      972                                     Jon Kabat-Zinn   
4      136                       John Eldredge/Stasi Eldredge   

                                               title  review  avg_ratings  
0      Arrows of the Queen (Heralds of Valdemar  #1)       2          5.0  
1  The Walking Dead  Book One (The Walking Dead #...       2          5.0  
2                                    Light in August       2          5.0  
3  Wherever You Go  There You Are: Mindfulness Me...       2          5.0  
4  Captivating: Unveiling the Mystery of a Woman'...       2          5.0

### Busco la editorial que ha publicado el mayor número de libros con más de 50 páginas.

In [ ]:
#Identifico la editorial que ha publicado el mayor número de libros con más de 50 páginas 

query= ''' 
               select publishers.publisher_id,
                      publishers.publisher,
                      count(books.title) as number_of_books
                from publishers
                  left join books on books.publisher_id = publishers.publisher_id
                where num_pages > 50
                group by publishers.publisher_id
                order by number_of_books desc
                limit 1
'''


ej_3 = pd.io.sql.read_sql(query, con = engine)
ej_3

publisher_id      publisher  number_of_books
0           212  Penguin Books               42

Podemos observar que la editorial que publicó el mayor número de libros con más de 50 páginas fue Penguin Books. Publicó un total de 42 libros.

### Identifico al autor que tiene la más alta calificación promedio : miro solo los libros con al menos 50 calificaciones.

In [ ]:
# Busco el autor que tiene la mayor calificación promedio

query= '''select l.author_id,
                 l.author,
                 round(avg(ratings.rating),2) as avg_rating                
          from(select books.author_id,
                      authors.author,
                      ratings.book_id,
                      count(ratings.rating) as number_of_ratings
               from ratings
                  left join books on books.book_id = ratings.book_id
                  inner join authors on books.author_id = authors.author_id
               group by ratings.book_id,
                        books.author_id,
                        authors.author
               having count(ratings.rating) >= 50) as l
             left join ratings on l.book_id = ratings.book_id
         group by  l.author_id,
                   l.author
         order by avg_rating desc
         limit 1
'''

ej_4 = pd.io.sql.read_sql(query, con = engine)
ej_4

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré        4.29

Las autoras que tienen la calificación promedio más alta son J.K. Rowling/Mary GrandPré. Recibieron una calificación promedio de 4.29.	

### Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [ ]:
# Busco número promedio de reseñas entre los usuarios que calificaron más de 50 libros

query= ''' select round(avg(m.text)) as avg_text_reviews   -- Saco promedio de reseñas
           from (select q.username,                          --Busco las reseñas realizadas por los usuarios que calificaron más de 50 veces.
                        count(reviews.text) as text
                 from (select username,                      --Busco usuarios que calificaron más de 50 libros
                              count(book_id) as graded_books
                       from ratings
                       group by username
                       having count(book_id) > 50) as q
                    left join reviews on reviews.username= q.username
                group by q.username) as m
'''


ej_5 = pd.io.sql.read_sql(query, con = engine)
ej_5

avg_text_reviews
0              24.0

El número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros es de 24 reseñas.

## Conclusiones

Para concluir, esta plataforma de servicios ofrece 1000 libros con 999 títulos diferentes (hay un libro llamado **Memoirs of a Geisha** que aparece en el servicio publicado por dos editoriales diferentes) , 636 autores y 340 editoriales diferentes.

Del total de libros, 819 fueron publicados después del 1 de enero de 2020. La editorial que publicó el mayor número de libros con más de 50 páginas fue Penguin Books con un total de 42 libros. Las autoras que tienen la calificación promedio más alta son J.K. Rowling/Mary GrandPré; recibieron una calificación promedio de 4.29. El número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros es de 24 reseñas.
